statefun-sdk-js/src/core.ts

```diff
@@ -0,0 +1,329 @@
+    get type() {
+        return this.#type;
+    }
+
+    get expireAfterWrite() {
+        return this.#expireAfterWrite;
+    }
+
+    get expireAfterCall() {
+        return this.#expireAfterCall;
+    }
+}
+
+
+/**
+ * An internal representation of a function spec.
+ * A function specification has a typename, a list of zero or more declared states, and an instance of a function to invoke.
+ */
+class FunctionSpec implements FunctionOpts {
+    readonly #typename;
+    readonly #fn;
+    readonly #valueSpecs;
+
+    constructor(typename: string, fn: (context: Context, message: Message) => void | Promise<void>, specs: ValueSpec[]) {
+        validateTypeName(typename);
+        if (fn === undefined) {
+            throw new Error(`input function must be defined.`);
+        }
+        this.#typename = typename;
+        this.#fn = fn;
+        this.#valueSpecs = specs;
+    }
+
+    static fromOpts({fn, specs, typename}: FunctionOpts): FunctionSpec {
+        validateTypeName(typename);
+        if (fn === undefined || fn === null) {
+            throw new Error(`missing function instance for ${typename}`);
+        }
+        let validatedSpecs = [];
+        let seen: Record<string, ValueSpec> = {};
+        for (let spec of (specs || [])) {
+            const valueSpec = ValueSpec.fromOpts(spec);
+            if (seen.hasOwnProperty(valueSpec.name)) {
+                throw new Error(`${valueSpec.name} is already defined.`);
+            }
+            seen[valueSpec.name] = valueSpec;
+            validatedSpecs.push(valueSpec);
+        }
+        return new FunctionSpec(typename, fn, validatedSpecs);
+    }
+
+    get valueSpecs() {
+        return this.#valueSpecs;
+    }
+
+    get fn() {
+        return this.#fn;
+    }
+
+    get typename() {
+        return this.#typename;
+    }
+}
+
+/**
+ *
+ * @param {string} typename a namespace/name string
+ */
+function validateTypeName(typename: string) {
+    parseTypeName(typename);
+}
+
+/**
+ * @param {string} typename a string of  <namespace>/<name>
+ * @returns {{namespace: string, name: string}}
+ */
+function parseTypeName(typename: string) {
+    if (isEmptyOrNull(typename)) {
+        throw new Error(`typename must be provided and of the form <namespace>/<name>`);
+    }
+    const index = typename.lastIndexOf("/");
+    if (index < 0 || index > typename.length) {
+        throw new Error(`Unable to find a / in ${typename}`);
+    }
+    const namespace = typename.substring(0, index);
+    const name = typename.substring(index + 1);
+    if (namespace === undefined || namespace.length === 0 || name === undefined || name.length === 0) {
+        throw new Error(`Illegal ${typename}, it must be of a form <namespace>/<name>`);
+    }
+    return {namespace, name};
+}
+
+function isEmptyOrNull(s: string | undefined | null) {
+    // noinspection SuspiciousTypeOfGuard
+    return (s === null || s === undefined || (typeof s !== 'string') || s.length === 0);
+}
+
+export {FunctionSpec}
```

Comment:
It'd be nicer to add the `export` directly at the types instead of doing all of this on the bottom.